File: lib/python/reference-peer.py
#!/usr/bin/env python3
import asyncio
import logging
import os
import socket
import sys

from the_mule import MuleClientBuilder


def local_ip() -> str:
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    except Exception:
        return "0.0.0.0"
    finally:
        s.close()


def extract_port(listen_addr: str) -> str:
    parts = listen_addr.split("/")
    for i, p in enumerate(parts):
        if p == "udp" and i + 1 < len(parts):
            return parts[i + 1]
    return "0"


async def main() -> None:
    client = await MuleClientBuilder().build()
    ip = local_ip()
    port = extract_port(os.environ.get("LISTEN_ADDR", ""))
    await client.send_status(f"started|/ip4/{ip}/udp/{port}/quic-v1")

    async for command in client:
        logging.info(f"received command: {command}")
        if command == "shutdown":
            await client.send_status("stopped")
            break
        elif command.startswith("restart|"):
            delay = command.split("|", 1)[1]
            await client.send_status("restarting")
            with open("/tmp/delay", "w") as f:
                f.write(delay)
            sys.exit(42)


if __name__ == "__main__":
    asyncio.run(main())

File: lib/rust/src/bin/reference-peer.rs
use tokio_stream::StreamExt;

fn local_ip() -> String {
    let socket = match std::net::UdpSocket::bind("0.0.0.0:0") {
        Ok(s) => s,
        Err(_) => return "0.0.0.0".to_string(),
    };
    if socket.connect("8.8.8.8:80").is_err() {
        return "0.0.0.0".to_string();
    }
    socket
        .local_addr()
        .map(|a| a.ip().to_string())
        .unwrap_or_else(|_| "0.0.0.0".to_string())
}

fn extract_port(listen_addr: &str) -> String {
    let parts: Vec<&str> = listen_addr.split('/').collect();
    for (i, p) in parts.iter().enumerate() {
        if *p == "udp" {
            if let Some(port) = parts.get(i + 1) {
                return port.to_string();
            }
        }
    }
    "0".to_string()
}

fn detect_multiaddr() -> String {
    let port = std::env::var("LISTEN_ADDR")
        .map(|a| extract_port(&a))
        .unwrap_or_else(|_| "0".to_string());
    let ip = local_ip();
    format!("/ip4/{ip}/udp/{port}/quic-v1")
}

#[tokio::main]
async fn main() {
    let mut client = the_mule::MuleClientBuilder::new()
        .build()
        .await
        .expect("failed to build mule client");

    let multiaddr = detect_multiaddr();
    client
        .send_status(&format!("started|{multiaddr}"))
        .await
        .expect("failed to send started");

    loop {
        let cmd = {
            use std::pin::Pin;
            let mut pinned = Pin::new(&mut client);
            pinned.next().await
        };
        match cmd {
            Some(the_mule::Command::Shutdown) => {
                let _ = client.send_status("stopped").await;
                break;
            }
            Some(the_mule::Command::Restart { delay_secs }) => {
                let _ = client.send_status("restarting").await;
                let _ = std::fs::write("/tmp/delay", delay_secs.to_string());
                std::process::exit(42);
            }
            Some(cmd) => {
                tracing::info!("received command: {:?}", cmd);
            }
            None => break,
        }
    }
}
